### PR TITLE
Fix data partitioning

### PR DESCRIPTION
**Pull request by daemon**

### CODE REVIEW

File: utils/model.py [hunk 1]
```diff
@@ -235,13 +236,16 @@ def splits(cls, config):
...
-                bucket = int(hashlib.sha1(filename.encode()).hexdigest(), 16) % 100
-                if bucket < train_pct:
-                    tag = 0
-                elif bucket < train_pct + dev_pct:
+                hashname = re.sub(r"_nohash_.*$", "", filename)
+                max_no_wavs = 2**27 - 1
+                bucket = int(hashlib.sha1(hashname.encode()).hexdigest(), 16)
+                bucket = (bucket % (max_no_wavs + 1)) * (100. / max_no_wavs)
+                if bucket < dev_pct:
                     tag = 1
```
Comment (tuzhucheng) [inline utils/model.py]: Should we use an Enum for these tags to specify what they mean?

**Comment by lintool** [comment]
Can we have a high-level comment on what this patch is designed to fix... a bit of broader context?

**Comment by daemon** [comment]
The train/dev/test datasets were being chosen incorrectly. Since the speech commands dataset contains speakers saying the same word multiple times, the train/dev/test datasets shouldn't share any of the same speakers; doing so leads to an overestimation of model accuracy (in theory). The dataset identifies speakers in the filename, which the TF impl uses to partition the datasets accordingly.

I thought this bug would explain why the PT port had done better than the TF impl, but it didn't. The PT port still does better than the TF impl.

File: utils/model.py (hunk 1 above)
Comment (daemon) [reply utils/model.py]: Sure

**Review by tuzhucheng — APPROVED** [review]

**Comment by lintool** [comment]
> Since the speech commands dataset contains speakers saying the same word multiple times, the train/dev/test datasets shouldn't share any of the same speakers

I don't see anything wrong with this, actually - technically, we would call this stratification by user: for each user, some of the examples are included in the training/dev/test. This tell us how well the system would do training on _my_ _own_ _voice_ - which is interesting.

Making sure the same user doesn't appear in the test is interesting also - it tells us how well we would do if the system never heard my name before...

Both are interesting test conditions... let's add an option to switch between the two?

**Comment by daemon** [comment]
Ah I see, I'll add an option.

**Pull request merged**